com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/CompletionsProvider.java

```diff
@@ -0,0 +1,118 @@
+/*******************************************************************************
+* Copyright (c) 2018 Microsoft Corporation and others.
+* All rights reserved. This program and the accompanying materials
+* are made available under the terms of the Eclipse Public License v1.0
+* which accompanies this distribution, and is available at
+* http://www.eclipse.org/legal/epl-v10.html
+*
+* Contributors:
+*     Microsoft Corporation - initial API and implementation
+*******************************************************************************/
+
+package com.microsoft.java.debug.plugin.internal;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.jdt.core.CompletionProposal;
+import org.eclipse.jdt.core.IType;
+import org.eclipse.jdt.core.search.IJavaSearchConstants;
+import org.eclipse.jdt.core.search.IJavaSearchScope;
+import org.eclipse.jdt.core.search.SearchEngine;
+import org.eclipse.jdt.core.search.SearchMatch;
+import org.eclipse.jdt.core.search.SearchParticipant;
+import org.eclipse.jdt.core.search.SearchPattern;
+import org.eclipse.jdt.core.search.SearchRequestor;
+import org.eclipse.jdt.ls.core.internal.contentassist.CompletionProposalRequestor;
+import org.eclipse.jdt.ls.core.internal.handlers.JsonRpcHelpers;
+
+import com.microsoft.java.debug.core.Configuration;
+import com.microsoft.java.debug.core.DebugException;
+import com.microsoft.java.debug.core.adapter.ICompletionsProvider;
+import com.microsoft.java.debug.core.protocol.Types.CompletionItem;
+import com.sun.jdi.StackFrame;
+
+public class CompletionsProvider implements ICompletionsProvider {
+
+    private static final Logger logger = Logger.getLogger(Configuration.LOGGER_NAME);
+
+    @Override
+    public List<CompletionItem> codeComplete(StackFrame frame, String snippet, int line, int column) {
+        List<CompletionItem> res = new ArrayList<CompletionItem>();
+
+        try {
+            IType type = resolveType(frame);
+            if (type != null) {
+                final int offset = JsonRpcHelpers.toOffset(type.getCompilationUnit().getBuffer(), frame.location().lineNumber(), 0);
```

Comment:
In order to keep the least dependency on jdt.ls, previously we always try to avoid using jdt.ls code in our implementation. Is it possible to keep the completeProvider implementation in jdt.ls, and then send the RPC request to jdt.ls for result?

Reply:
The main implementation is different from the jdt.ls.  The JDT is code complete on compilation unit with offset without *snippet* while our implementation is based on *snippet*. 

Here we leverage the CompletionProposalRequestor from the JDT.LS. Otherwise, it just copy this class to the debugger side which is not necessary.